funnel/metadata_db/main.py

```diff
@@ -124,6 +125,106 @@ def __contains__(self, key: str) -> bool:
         return key in self._df.index
 
 
+@pydantic.dataclasses.dataclass
+class CacheMetadataStore(BaseMetadataStore):
+    """Uses the cache_store as a database"""
+
+    readonly: bool = False
+    serializer: str = 'auto'
+    serializer_dump_kwargs: typing.Dict[str, typing.Any] = pydantic.Field(default_factory=dict)
+    serializer_load_kwargs: typing.Dict[str, typing.Any] = pydantic.Field(default_factory=dict)
+
+    def __post_init_post_parse__(self):
+        self.cache_store._ensure_dir(self._metadata_store_path)
+
+    @property
+    def _metadata_store_path(self) -> str:
+        return self.cache_store._construct_item_path('funnel_metadata_store')
+
+    def _artifact_meta_path(self, key: str) -> str:
+        return f'{self._metadata_store_path}/{key}.artifact.json'
+
+    def put(
+        self, key: str, value, serializer: str = 'auto', dump_kwargs: typing.Optional[dict] = None
+    ) -> None:
+        """Records and serializes key with its corresponding value in the metadata and cache store.
+
+        Parameters
+        ----------
+        key : str
+
+        value : typing.Any
+            Any serializable Python object
+        serializer : str
+            The name of the serializer you want to use. The built-in
+            serializers are:
+
+                - 'auto' (default): automatically choose the serializer based on the type of the value
+                - 'xarray.netcdf': requires xarray and netCDF4
+                - 'xarray.zarr': requires xarray and zarr
+
+            You can also register your own serializer via the @funnel.registry.serializers.register decorator.
+        dump_kwargs : dict
+            Additional keyword arguments to pass to the serializer when dumping artifact to the cache store.
+        """
+
+        serializer = serializer or self.serializer
+        dump_kwargs = dump_kwargs or self.serializer_dump_kwargs
+        artifact = self.cache_store.put(key, value, serializer, dump_kwargs=dump_kwargs)
+        with self.cache_store.fs.open(self._artifact_meta_path(key), 'w') as fobj:
+            fobj.write(artifact.json(indent=2))
+
+    def get(self, key: str, **load_kwargs) -> typing.Any:
+        """Returns the value for the key if the key is in both the metadata and cache stores.
+
+        Parameters
+        ----------
+        key : str
+        load_kwargs : dict
+            Additional keyword arguments to pass to the serializer when loading artifact from the cache store.
+
+        Returns
+        -------
+        value :
+            the value for the key if the key is in both the metadata and cache stores.
+        """
+        # TODO: Discuss whether the dataframe is overkill here or not
+        # note: in an effort to avoid using a database/dataframe, I've made two primary simplifications here:
+        # 1. I don't use the artifacts predetermine load_kwargs
+        # 2. I've hard coded the serialzer
```

Comment:
After thinking on this for a few days, I've come to think the dataframe is indeed overkill.  We want to cache metadata once its been read by funnel but, we don't need to read all metadata records up front. I think a lazy dict interface for the metadata may make the most sense:

```python
from collections.abc import MutableMapping
import json


class LazyMetadataDict(MutableMapping):

    def __init__(self, cache_store, *args, **kwargs):
        self._cache_store = cache_store
        self._metadata = dict(*args, **kwargs)

    def __contains__(self, key: str) -> bool:
        return key in self._metadata or key in self._cache_store.mapper

    def __getitem__(self, key):
        if key not in self._metadata:
            # raise key error if not in cache store
            self._metadata[key] = json.loads(self._cache_store.mapper[key])
        return self._metadata[key]
```

using this instead of `df`, we could still check for metadata keys and get json objects back, but wouldn't have to eagerly load them.

Reply:
This looks neat. I'm currently looking into it

Reply:
I was able to get the job done without The `LazyMetadataDict`. Also, I was able to consolidate everything under a single object `CacheStore`